### PR TITLE
[#283] Add docs for Custom combinators

### PR DESCRIPTION
**Pull request by vrom911**

### CODE REVIEW

File: src/Toml/Codec/Combinator/Custom.hs [hunk 1]
```diff
@@ -32,20 +34,111 @@ import Toml.Type.AnyValue (AnyValue)
...
--- | Codec for text values with custom error messages for parsing.
+{- | Codec for text values with custom error messages for parsing.
+
+__Example:__
+
+We have the following type that represents the image format:
+
+@
+__data__ Format
+    | Jpeg
```
Comment (chshersh) [inline src/Toml/Codec/Combinator/Custom.hs:45]: Small typo
```suggestion
    = Jpeg
```

File: src/Toml/Codec/Combinator/Custom.hs [hunk 2]
```diff
@@ -32,20 +34,111 @@ import Toml.Type.AnyValue (AnyValue)
...
--- | Codec for values with a 'Read' and 'Show' instance.
+{- | Codec for values with a 'Read' and 'Show' instance.
+
+__Example:__
+
+We have the following type that represents the image format:
+
+@
+__data__ Format
+    | Jpeg
```
Comment (chshersh) [inline src/Toml/Codec/Combinator/Custom.hs:104]:
```suggestion
    = Jpeg
```

File: src/Toml/Codec/Combinator/Custom.hs [hunk 3]
```diff
@@ -32,20 +34,111 @@ import Toml.Type.AnyValue (AnyValue)
...
 {-# INLINE read #-}
 
 {- | Codec for general nullary sum data types with a 'Bounded', 'Enum', and
-'Show' instance. This codec provides much better error messages than 'read' for
-nullary sum types.
+'Show' instance. This codec is similar to 'read' but provides much better error
+messages than 'read' for nullary sum types.
+
+E.g. for the same @Format@ example from 'read' function, but with the
+@Toml.'enumBounded' "foo"@ codec the error for @foo = "jif"@ in the @TOML@ file
```
Comment (chshersh) [inline src/Toml/Codec/Combinator/Custom.hs:136]: To be consistent with the error message below.
```suggestion
@Toml.'enumBounded' "foo"@ codec the error for @foo = "Jif"@ in the @TOML@ file
```

File: src/Toml/Codec/Combinator.hs [hunk 4]
```diff
@@ -93,3 +94,13 @@ import Toml.Codec.Combinator.Set (arrayHashSetOf, arrayIntSet, arraySetOf, hashS
...
+
+{- $custom
+This module provides additional combinators that could help in the situation
+when some additional manipulations for the standard combinators is required.
+
+/For example,/ 'validate' allows to perform some custom validation on the codec
+before encoding. And 'enumBounded' is an automatical codec that uses 'Enum' and
+'Bounded' instances of the data type only and provides descriptive error
+messages at the same time.
+-}
```
Comment (chshersh) [inline src/Toml/Codec/Combinator.hs:106]: Very nice section description!

File: src/Toml/Codec/Combinator/Monoid.hs [hunk 5]
```diff
@@ -70,28 +70,32 @@ Decodes to @'Any' 'False'@ when the key is not present.
...
 {- | Codec for 'Product' wrapper for given converter's values.
 
+Decodes to @'Product' 1@ when the key is not present.
+
 @since 1.2.1.0
 -}
-product :: (Key -> TomlCodec a) -> Key -> TomlCodec (Product a)
-product codec = diwrap . codec
+product :: (Num a) => (Key -> TomlCodec a) -> Key -> TomlCodec (Product a)
+product codec = dimap (Just . getProduct) (maybe mempty Product) . dioptional . codec
```
Comment (chshersh) [inline src/Toml/Codec/Combinator/Monoid.hs:93]: This module just became much cooler! Thanks :+1:

File: examples/biTest.toml [hunk 6]
```diff
@@ -4,13 +4,13 @@ testF = 1.23
...
 
-testEven = 5
+testEven = 6
 # testEven = 5
 
 testN = "newtype"
 # testStatus = "username"
 
-testC = "Light"
+testC = "Light+"
```
Comment (vrom911) [inline examples/biTest.toml:13]:
```suggestion
testC = "Light"
```

**Pull request merged**